feat(assertions): an initial version of the CDK 'assert' library, now experimentally available in all supported languages (#14952)
aws/aws-cdk: packages/@aws-cdk/assertions/README.md

# Assert v2
<!--BEGIN STABILITY BANNER-->

---

![cdk-constructs: Experimental](https://img.shields.io/badge/cdk--constructs-experimental-important.svg?style=for-the-badge)

> The APIs of higher level constructs in this module are experimental and under active development.
> They are subject to non-backward compatible changes or removal in any future version. These are
> not subject to the [Semantic Versioning](https://semver.org/) model and breaking changes will be
> announced in the release notes. This means that while you may use them, you may need to update
> your source code when upgrading to a newer version of this package.

---

<!--END STABILITY BANNER-->

> This module allows asserting the contents of CloudFormation templates.

Reviewer: I think that in the future we might also want to support creating assertions against things like cloud assemblies, assets, etc, so I wouldn't want to scope this down too much at the outset:

```suggestion
Functions for writing test asserting against CDK applications, with focus on CloudFormation templates.
```

> To run assertions based on a CDK `Stack`, start off with -

Reviewer: ```suggestion
The `TemplateAssertions` class includes a set of methods for writing assertions against CloudFormation templates. Use one of the `TemplateAssertions.fromXxx()` static methods to create an instance of this class.

To create `TemplateAssertions` from CDK stack, start off with:
```

```ts
import { Stack } from '@aws-cdk/core';
import { TemplateAssertions } from '@aws-cdk/assertions';

const stack = new Stack(...)
...
const assert = TemplateAssertions.fromStack(stack);
```

Alternatively, assertions can be run on an existing CloudFormation template -

```ts
const template = fs.readFileSync('/path/to/template/file');
const assert = TemplateAssertions.fromTemplate(template);
```

## Full Template Match

The simplest assertion would be to assert that the template matches a given
template.

```ts

> // In typescript

Reviewer: ```suggestion
```

assert.templateMatches({
  Resources: {
    Type: 'Foo::Bar',
    Properties: {
      Baz: 'Qux',
    },
  },
});
```

## Counting Resources

This module allows asserting the number of resources of a specific type found
in a template.

```ts
assert.resourceCountIs('Foo::Bar', 2);
```

## Resource Matching

Beyond resource counting, the module also allows asserting that a resource with
specific properties are present.

The following code asserts that the `Properties` section of a resource of type
`Foo::Bar` contains the specified properties -

```ts

> assert.hasResource('Foo::Bar', {

Reviewer: How about we call this `hasResourceProperties` and the `COMPLETE` version `hasResourceDefinition`? The API in which you pass `{ part: ResourcePart.COMPLETE }` is pretty awkward.

Author: I like it!
